src/opamp.rs

```diff
@@ -28,81 +30,229 @@
+    output: PhantomData<Output>,
+}
+
+/// Represents the output when the opamp is connected to the
+/// ADC internally, and not connected to a pin.
+pub struct InternalOutput;
+
+/// Trait for opamps that can be run in follower mode.
+pub trait IntoFollower<Opamp, IntoInput, IntoOutput, Input, Output> {
+    /// Coonfigures the opamp as voltage follower.
+    fn follower(self, input: IntoInput, output: IntoOutput) -> Follower<Opamp, Input, Output>;
+}
+/// Trait for opamps that can be run in open-loop mode.
+pub trait IntoOpenLoop<
+    Opamp,
+    IntoNonInverting,
+    IntoInverting,
+    IntoOutput,
+    NonInverting,
+    Inverting,
+    Output,
+>
+{
+    /// Configures the opamp for open-loop operation.
+    fn open_loop(
+        self,
+        non_inverting: IntoNonInverting,
+        inverting: IntoInverting,
+        output: IntoOutput,
+    ) -> OpenLoop<Opamp, NonInverting, Inverting, Output>;
+}
+/// Trait for opamps that can be run in programmable gain mode.
+#[allow(private_bounds)]
+pub trait IntoPga<Opamp, NonInverting, Output>
+where
+    Opamp: ConfigurePgaReg<Opamp, NonInverting> + LookupPgaGain,
+{
+    /// Configures the opamp for programmable gain operation.
+    fn pga<B: Borrow<NonInverting>>(
+        self,
+        non_inverting: B,
+        output: Output,
+        gain: Gain,
+    ) -> Pga<Opamp, NonInverting, Output>;
+
+    /// Trait for opamps that can be run in programmable gain mode,
+    /// with external filtering.
+    fn pga_external_filter<
+        B1: Borrow<NonInverting>,
+        B2: Borrow<<Opamp as ConfigurePgaReg<Opamp, NonInverting>>::vinm0>,
+    >(
+        self,
+        non_inverting: B1,
+        filter: B2,
+        output: Output,
+        gain: Gain,
+    ) -> Pga<Opamp, NonInverting, Output>;
+
+    /// Configures the opamp for programmable gain operation, with
+    /// external filtering.
+    fn pga_external_bias<
+        B1: Borrow<NonInverting>,
+        B2: Borrow<<Opamp as ConfigurePgaReg<Opamp, NonInverting>>::vinm0>,
+    >(
+        self,
+        non_inverting: B1,
+        inverting: B2,
+        output: Output,
+        gain: Gain,
+    ) -> Pga<Opamp, NonInverting, Output>;
+
+    /// Configures the opamp for programmable gain operation, with
+    /// external filtering.
+    fn pga_external_bias_and_filter<
+        B1: Borrow<NonInverting>,
+        B2: Borrow<<Opamp as ConfigurePgaReg<Opamp, NonInverting>>::vinm0>,
+        B3: Borrow<<Opamp as ConfigurePgaReg<Opamp, NonInverting>>::vinm1>,
+    >(
+        self,
+        non_inverting: B1,
+        inverting: B2,
+        filter: B3,
+        output: Output,
+        gain: Gain,
+    ) -> Pga<Opamp, NonInverting, Output>;
+}
+
+/// Internal trait implementing the low level register write used to
+/// configure the PGA.
+trait ConfigurePgaReg<Opamp, NonInverting>
+where
+    Opamp: LookupPgaGain,
+{
+    /// Type of the associated vinm0 input.
+    type vinm0;
+    /// Type of the associated vinm1 input.
+    type vinm1;
```

Comment:
CI seems unhappy with these names.

What about something like this? :)

```suggestion
    /// Type of the associated vinm0 input.
    type Vinm0;
    /// Type of the associated vinm1 input.
    type Vinm1;
```

Reply:
Thanks @usbalbin  - 489c268

I don't have the hardware set up to test this any more, and likely won't be able to get back to it in the near future, but it was working well when I put it down.